torchsde/_core/adjoint_sde.py

```diff
@@ -0,0 +1,425 @@
+            )
+            extra_vjp_y_and_params = misc.grad(
+                outputs=g_eval,
+                inputs=y + params,
+                grad_outputs=adj_times_dgdx,
+                allow_unused=True,
+                create_graph=True,
+            )
+            extra_vjp_y, extra_vjp_params = extra_vjp_y_and_params[:n_tensors], extra_vjp_y_and_params[n_tensors:]
+            extra_vjp_params = misc.flatten(extra_vjp_params)
+
+            # Vector field change due to log-ratio term, i.e. ||u||^2 / 2.
+            h_eval = sde.h(-t, y)
+            u_eval = misc.seq_sub_div(f_eval, h_eval, g_eval)
+            log_ratio_correction = [.5 * torch.sum(u_eval_ ** 2., dim=1) for u_eval_ in u_eval]
+            corr_vjp_y_and_params = misc.grad(
+                outputs=log_ratio_correction,
+                inputs=y + params,
+                grad_outputs=adj_l,
+                allow_unused=True,
+            )
+            corr_vjp_y, corr_vjp_params = corr_vjp_y_and_params[:n_tensors], corr_vjp_y_and_params[n_tensors:]
+            corr_vjp_params = misc.flatten(corr_vjp_params)
+
+            vjp_y = misc.seq_add(vjp_y, extra_vjp_y, corr_vjp_y)
+            vjp_params = vjp_params + extra_vjp_params + corr_vjp_params
+
+        return (*f_eval_corrected, *vjp_y, *vjp_l, vjp_params)
+
+    ########################################
+    #             g_prod_logqp             #
+    ########################################
+
+    def g_prod_default_logqp(self, t, y_aug, v):
+        n_tensors = self._n_tensors
+        adj_l = y_aug[2 * n_tensors:3 * n_tensors]
+        vjp_l = [torch.zeros_like(adj_l_) for adj_l_ in adj_l]
+        results = self.g_prod_default(t, y_aug, v)
+        g_prod_eval, vjp_y, vjp_params = results[:n_tensors], results[n_tensors:2 * n_tensors], results[2 * n_tensors]
+        return (*g_prod_eval, *vjp_y, *vjp_l, vjp_params)
+
+    def g_prod_diagonal_logqp(self, t, y_aug, v):
+        n_tensors = self._n_tensors
+        adj_l = y_aug[2 * n_tensors:3 * n_tensors]
+        vjp_l = [torch.zeros_like(adj_l_) for adj_l_ in adj_l]
+        results = self.g_prod_diagonal(t, y_aug, v)
+        g_prod_eval, vjp_y, vjp_params = results[:n_tensors], results[n_tensors:2 * n_tensors], results[2 * n_tensors]
+        return (*g_prod_eval, *vjp_y, *vjp_l, vjp_params)
+
+    ########################################
+    #             gdg_prod_logqp           #
+    ########################################
+
+    def gdg_prod_diagonal_logqp(self, t, y_aug, v):
+        n_tensors = self._n_tensors
+        adj_l = y_aug[2 * n_tensors:3 * n_tensors]
+        vjp_l = [torch.zeros_like(adj_l_) for adj_l_ in adj_l]
+        results = self.gdg_prod_diagonal(t, y_aug, v)
+        gdg_v, vjp_y, vjp_params = results[:n_tensors], results[n_tensors:2 * n_tensors], results[2 * n_tensors]
+        return (*gdg_v, *vjp_y, *vjp_l, vjp_params)
+
+    def gdg_prod_default_logqp(self, t, y_aug, v):
+        raise NotImplementedError
```

Comment:
The Strat adjoint should be quite clear as the low-order solvers uses `f_uncorrected` and `g_prod_default`. The drift term is basically what you get in ODEs. The diffusion term is slightly more tricky, but if you write the forward SDE's diffusion out as a weighted (by the m Brownian dimensions) sum of diffusion columns, it could be easily derived as m ODE terms. 

The Ito stuff is a bit hairy, and it took me quite a long while to derive those. There's virtually no real way to actually check this apart from running rigorous numerical tests. Thanks for putting in the work of going through the code nevertheless!